### PR TITLE
Feature/tdd

### PR DESCRIPTION
**Pull request by citec-47**
In this PR, we did the testing using the TDD method with the following requirements 📑

General requirements
✅ Make sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
✅ Make sure that you used correct [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
✅ Make sure that you documented your work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).

Ruby requirements
✅ Follow our list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).

Project requirements
✅ Create a class called Solver.
✅ Create a method called [factorial](https://en.wikipedia.org/wiki/Factorial) that takes one argument, an integer N, and returns the factorial for that number. The factorial is the multiplication of all integers from 1 to N and has the special case that the factorial of 0 is 1. This method only accepts 0 and positive integers, so if a negative integer is given it should raise an exception.
✅ Create a method called reverse that takes one argument, a string word, and returns word reversed (e.g. if word is "hello" it returns "olleh").

Create a method called fizzbuzz that takes one argument, an integer N, and returns a string. The returned string is constructed following these rules:
✅ When N is divisible by 3, return "fizz".
✅ When N is divisible by 5, return "buzz".
✅ When N is divisible by 3 and 5, return "fizzbuzz".
✅ Any other case, return N as a string (e.g. say N is 7 then return "7").
✅ Remember to start with tests for all of the methods described above. Your commit history should prove that you have used TDD.

### CODE REVIEW

File: spec/tdd_solver_spec.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+describe Solver do
+  before(:each) do
+    @solver = Solver.new
+  end
+
+  context 'testing the factorial method' do
+    it 'returns 1 when given 0' do
+      expect(@solver.factorial(0)).to eql 1
+    end
+
```
Comment (t-yanick) [inline spec/tdd_solver_spec.rb:12]: - [ ]Great job with the tests cases. Your are supposed to have a test case for factorial that raises an exception when a negative number is given as per project requirements.
image

File: spec/tdd_solver_spec.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      expect(@solver.fizzbuzz(6)).to eql 'fizz'
+    end
+
+    it 'returns "buzz" when divisible by 5' do
+      expect(@solver.fizzbuzz(10)).to eql 'buzz'
+    end
+
+    it 'returns "fizzbuzz" when divisible by 3 and 5' do
+      expect(@solver.fizzbuzz(15)).to eql 'fizzbuzz'
+    end
```
Comment (t-yanick) [inline spec/tdd_solver_spec.rb:39]: - [ ]In testing the fizzbuzz method, you have missed one test case: `Any other case, return N as a string (e.g. say N is 7 then return "7").` that is **N is not divisible by 3 or 5**

**Review by t-yanick — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED

Hi @citec-47 and @zunairkhan811 , 

### Good job so far!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Collaborative work :+1: 
:heavy_check_mark: TDD respected. 
:heavy_check_mark: Use of correct Gitflow
:heavy_check_mark: All added test cases passing

There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️

- [ ] Kindly ensure that you make your commit messages complete sentences in the imperative tense and always capitalise the first letter.
- [ ] Also use descriptive names for your branches. **feature** is not a good branch name. It does not indicate what has been implemented in that branch and in big codebases it can be a problem

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/tdd_solver_spec.rb (hunk 1 above)
Comment (zunairkhan811) [reply spec/tdd_solver_spec.rb:12]: Thanks! The test case have been added

File: spec/tdd_solver_spec.rb (hunk 2 above)
Comment (zunairkhan811) [reply spec/tdd_solver_spec.rb:39]: Thanks! The test case have been added

**Review by Deepakdanger — APPROVED** [review]
Hi @citec-47 & @zunairkhan811 , 

### Status: Approved ✔️

Your project is complete!  There is nothing else to say other than... it's time to merge it  🚢 
Congratulations! 🎉 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- x

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**